### PR TITLE
Verify important QR code attributes on scan (EXPOSUREAPP-6378/6399)

### PR DESCRIPTION
**Pull request by kolyaopahle**
### Description
This PR adds validation for QR Code content to the checkin process

According to the spec we currently check:
1. if a description is set and not empty
2. if an address is set and not empty
3. if the cryptographic Seed is equal to 16 bytes in length

### Steps to reproduce
1. Try scanning a valid qr code
2. checkin should succeed
3. Try scanning qr codes that had their parameters destroyed
4. checkin should fail with an error dialog

### CODE REVIEW

**Review by janetback — DISMISSED** [review]
UA reviewed.

**Comment by harambasicluka** [comment]
Empty description
image

Empty address
image

Over 100 char description
image

Over 100 char address
image

Address with new line
image

Description with new line
image

Start larger than end time
image

17Byte cryptographicSeed
image

**Review by harambasicluka — DISMISSED** [review]

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/eventregistration/checkins/qrcode/TraceLocationVerifier.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
+package de.rki.coronawarnapp.eventregistration.checkins.qrcode
+
+import androidx.annotation.StringRes
+import dagger.Reusable
+import de.rki.coronawarnapp.R
+import de.rki.coronawarnapp.server.protocols.internal.pt.TraceLocationOuterClass
+import javax.inject.Inject
+
+@Reusable
+class TraceLocationVerifier @Inject constructor() {
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/eventregistration/checkins/qrcode/TraceLocationVerifier.kt:10]: Please add some unit tests for it ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/eventregistration/checkins/qrcode/TraceLocationVerifier.kt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            object InvalidStartEndTime : Invalid(R.string.trace_location_checkins_qr_code_invalid_times)
+            object InvalidCryptographicSeed :
+                Invalid(R.string.trace_location_checkins_qr_code_invalid_cryptographic_seed)
+        }
+    }
+
+    companion object {
+        const val CROWD_NOTIFIER_CRYPTO_SEED_LENGTH = 16
+        const val QR_CODE_DESCRIPTION_MAX_LENGTH = 100
+        const val QR_CODE_ADDRESS_MAX_LENGTH = 100
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/eventregistration/checkins/qrcode/TraceLocationVerifier.kt:74]: Those should be private , if you don't use them publicly

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/checkins/CheckInsFragment.kt [hunk 3]
```diff
@@ -125,6 +127,13 @@ class CheckInsFragment : Fragment(R.layout.trace_location_attendee_checkins_frag
         }
     }
 
+    private fun showInvalidQrCodeInformation(@StringRes errorTextRes: Int) =
+        AlertDialog.Builder(requireContext()).apply {
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/checkins/CheckInsFragment.kt:131]: Please MaterialBuilder?

Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/checkins/CheckInsFragment.kt:131]: Should we also change this for the other AlertBuilders in this Fragment?

**Comment by harambasicluka** [comment]
@kolyaopahle you can remove the `do not merge` label when the latest tech spec changes are implemented.

**Review by harambasicluka — APPROVED** [review]
## Valid QR Code

Start = End
image

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/checkins/CheckInsFragment.kt (hunk 3 above)
Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/attendee/checkins/CheckInsFragment.kt]: That would be great 👍

**Review by mtwalli — APPROVED** [review]
Thank you! 🙏

**Pull request merged**